controllers/phdeployment_controller.go

```diff
@@ -17,34 +17,587 @@ package controllers
+	Scheme        *runtime.Scheme
+	GraphqlClient *graphql.GraphqlClient
+	Ingress       PhIngress
 }
 
 // +kubebuilder:rbac:groups=primehub.io,resources=phdeployments,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=primehub.io,resources=phdeployments/status,verbs=get;update;patch
 
 func (r *PhDeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("phdeployment", req.NamespacedName)
+	ctx := context.Background()
+	log := r.Log.WithValues("phdeployment", req.NamespacedName)
+	seldonDeploymentKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}
+	ingressKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      "deploy-" + req.Name,
+	}
+
+	phDeployment := &primehubv1alpha1.PhDeployment{}
+	if err := r.Get(ctx, req.NamespacedName, phDeployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("PhDeployment deleted")
+		} else {
+			log.Error(err, "Unable to fetch PhShceduleJob")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Start Reconcile PhDeployment")
+	startTime := time.Now()
+	defer func() {
+		log.Info("Finished Reconciling phDeployment ", "phDeployment", phDeployment.Name, "ReconcileTime", time.Since(startTime))
+	}()
+
+	oldStatus := phDeployment.Status.DeepCopy()
+	phDeployment = phDeployment.DeepCopy()
+
+	if phDeployment.Status.History == nil {
+		phDeployment.Status.History = make([]primehubv1alpha1.PhDeploymentHistory, 0)
+	}
+
+	// update history
+	hasChanged := r.updateHistory(ctx, phDeployment)
+
+	// phDeployment has been stoped
+	if phDeployment.Spec.Stop == true {
+		// delete seldonDeployment
+		if err := r.deleteSeldonDeployment(ctx, seldonDeploymentKey); err != nil {
+			log.Error(err, "failed to delete seldonDeployment and stop phDeployment")
+			return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+		}
+
+		// delete ingress
+		// if err := r.deleteIngress(ctx, ingressKey); err != nil {
+		// 	log.Error(err, "failed to delete ingress and stop phDeployment")
+		// 	return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+		// }
+
+		// update stauts
+		phDeployment.Status.Phase = primehubv1alpha1.DeploymentStopped
+		phDeployment.Status.Messsage = "deployment has been stopped"
+		phDeployment.Status.Replicas = phDeployment.Spec.Predictors[0].Replicas
+		phDeployment.Status.AvailableReplicas = 0
+		phDeployment.Status.Endpoint = "" // TODO: should be hard coded
+
+		// update history
+		r.updateHistory(ctx, phDeployment)
+
+		if !apiequality.Semantic.DeepEqual(oldStatus, phDeployment.Status) {
+			if err := r.updatePhDeploymentStatus(ctx, phDeployment); err != nil {
+				return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// reconcile seldonDeployment
+	if err := r.reconcileSeldonDeployment(ctx, phDeployment, seldonDeploymentKey, hasChanged); err != nil {
+		log.Error(err, "reconcile Seldon Deployment error.")
```

Comment:
It prompt error when create a phdeployment
![image](https://user-images.githubusercontent.com/860633/78543638-8647ca00-782b-11ea-90a7-cd65796ccb86.png)

Reply:
fixed